feat(training): add weighted domain mixtures (#3303)
NVIDIA-NeMo/Automodel: docs/guides/llm/pretraining.mdx

Reviewer: ```suggestion
NeMo AutoModel provides an end-to-end recipe to run LLM pretraining with Hugging Face–native models and Megatron Core-style datasets.
```

Reviewer: ```suggestion
This guide explains how to pretrain OpenAI’s `GPT2-124M` model on a FineWeb-Edu subset of 10 billion tokens.
```

Reviewer: ```suggestion
[FineWeb-Edu](https://huggingface.co/datasets/HuggingFaceFW/fineweb-edu) is a dataset consisting of 1.3T tokens of educational web pages filtered from the larger [FineWeb](https://huggingface.co/datasets/HuggingFaceFW/fineweb) dataset. The educational web pages were filtered from the main dataset using a fine-tuned [BERT](https://huggingface.co/docs/transformers/en/model_doc/bert)-like classifier. You can find more information about the filtering process in the [FineWeb-Edu blog post](https://huggingface.co/spaces/HuggingFaceFW/blogpost-fineweb-v1).
```

Reviewer: ```suggestion
The following example shows the data format:
```

Reviewer: ```suggestion
This guide uses the FineWeb-Edu 10BT sample, which is a collection of approximately 10 billion tokens randomly drawn from the full FineWeb-Edu dataset. To prepare the data, run the following commands:
```

Reviewer: ```suggestion

```bash
```

Reviewer: ```suggestion

NeMo AutoModel provides a tool that tokenizes data and saves it in the Megatron Core dataset format. Use the tool as follows:
```

Reviewer: ```suggestion

```bash
```

Reviewer: ```suggestion
Set `--workers` to the number of CPU cores to use for parallel tokenization.
```

Reviewer: ```suggestion
- `setup()`: Builds all training components from the config.
```

Reviewer: ```suggestion
- `run_train_validation_loop()`: Executes training and validation steps.
```

Reviewer: ```suggestion
To add weights to the dataset blends, pass a list. For example, `paths: ["30", "fineweb_edu/megatron_gpt2/processed_data_0_text_document", "70", "fineweb_edu/megatron_gpt2/processed_data_1_text_document"]`.
```

Reviewer: ```suggestion
In these scenarios, you can pass `is_meta_device: true` in the model config. The model is then instantiated using [PyTorch's Meta device](https://docs.pytorch.org/docs/stable/meta.html), which loads no data but stores the parameter metadata necessary for sharding the model. After the model is sharded, only the weights required by each model shard are loaded.
```

Reviewer: ```suggestion
If you saved or plan to use the provided configuration at `examples/llm_pretrain/megatron_pretrain_gpt2.yaml`, run the following command:
```

Reviewer: ```suggestion
During training, stepwise logs report loss, memory usage, and tokens per second. The recipe saves checkpoints under the configured `checkpoints/` directory.
```

Reviewer: ```suggestion
As training progresses, you should observe the model loss beginning to converge. To verify your results, compare your convergence curves with the baseline benchmarks in the [llm.c repository](https://github.com/karpathy/llm.c/discussions/481).
```

Reviewer: ```suggestion
This example demonstrates how to pretrain an LLM by using the NeMo AutoModel library. It uses the LLM [training recipe](https://github.com/NVIDIA-NeMo/Automodel/blob/main/nemo_automodel/recipes/llm/train_ft.py), specifically `TrainFinetuneRecipeForNextTokenPrediction`. This recipe orchestrates loading, dataset preparation, optimizer setup, distributed training, checkpointing, and logging.
```

Reviewer: ```suggestion
A recipe in NeMo AutoModel is a self-contained orchestration module that wires together all
components needed to perform a specific task, such as pretraining.
It is equivalent to a `Trainer` class but is highly modular, stateful, and reproducible.
```

Reviewer: ```suggestion

<Tip>
```

Reviewer: ```suggestion
The following is a complete configuration based on `examples/llm_pretrain/megatron_pretrain_gpt2.yaml`:
```

  # Use the validation settings from the complete recipe above.

checkpoint:
  best_metric_key: weighted
```

For a domain with normalized sampling weight \(q_i\) and objective weight
\(w_i\), the recipe multiplies each supervised token loss by \(w_i / q_i\).
This keeps one optimizer and learning-rate schedule while making the expected
training loss equal to the requested weighted domain objective. The recipe
logs achieved supervised-token fractions for each domain, reports every
per-domain validation loss, and reports their objective-weighted aggregate as
`weighted`.

The `sampling_weight` values must match the explicit weights in
the `MegatronPretraining` dataset's `dataset.paths`, and domain order must
match the blend order. JSON blend files and implicit unweighted blends are not
supported. Sequence packing is not supported with `domain_mixture` because
packed batches do not currently preserve per-token domain IDs. Pipeline
parallelism is also not currently supported because its target contract does
not carry per-token objective weights. Context parallelism is disabled until
its distributed gradient contract is covered by parity tests. The configured

> loss must use `reduction: sum`; the built-in masked, fused-linear, chunked, and

Reviewer: ```suggestion
loss must use `reduction: sum`. The built-in masked, fused-linear, chunked, and
```

Transformer Engine parallel cross-entropy losses accept these objective
weights.

## Load Large Models

> In distributed training, the typical model-loading pipeline has each GPU load the entire model and then retain only the shard it needs. This approach becomes problematic when the model size exceeds the memory capacity of a single GPU. For instance, a 70B-parameter model requires about 140GB of memory for its parameters when using the BF16 data type (2 bytes per parameter). Since most widely used GPUs are limited to 80GB, the full model cannot be directly loaded onto a single device.

Reviewer: ```suggestion

In distributed training, the typical model-loading pipeline has each GPU load the entire model and then retain only the shard it needs. This approach becomes problematic when the model size exceeds the memory capacity of a single GPU. For example, a 70B-parameter model requires about 140 GB of memory for its parameters when using the BF16 data type (2 bytes per parameter). Because most widely used GPUs are limited to 80 GB, the full model cannot be loaded directly onto a single device.
```
